discovery/consul/consul.go

```diff
@@ -226,92 +286,138 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 		default:
 		}
 
+		// Get the local datacenter first, if necessary.
+		err := d.getDatacenter()
 		if err != nil {
-			level.Error(d.logger).Log("msg", "Error refreshing service list", "err", err)
-			rpcFailuresCount.Inc()
 			time.Sleep(retryInterval)
 			continue
 		}
-		// If the index equals the previous one, the watch timed out with no update.
-		if meta.LastIndex == lastIndex {
-			continue
-		}
-		lastIndex = meta.LastIndex
 
-		// If the datacenter was not set from clientConf, let's get it from the local Consul agent
-		// (Consul default is to use local node's datacenter if one isn't given for a query).
-		if d.clientDatacenter == "" {
-			info, err := d.client.Agent().Self()
-			if err != nil {
-				level.Error(d.logger).Log("msg", "Error retrieving datacenter name", "err", err)
-				time.Sleep(retryInterval)
-				continue
+		if len(d.watchedServices) == 0 || d.watchedTag != "" {
+			// We need to watch the catalog.
+			d.watchServices(ctx, ch, &lastIndex, services)
+		} else {
+			// We only have fully defined services.
+			for _, name := range d.watchedServices {
+				ctx, _ := context.WithCancel(ctx)
```

Comment:
do we need the cancel func here? if not we can just remove this line, seems like we're creating a new cancel func for each service in watchServices

Reply:
you're right, this isn't needed, the same context can be reused. Removing this.